File: src/main/java/gruvexp/bbminigames/mechanics/SteamPipe.java
package gruvexp.bbminigames.mechanics;

import gruvexp.bbminigames.Main;
import gruvexp.bbminigames.Util;
import org.bukkit.Axis;
import org.bukkit.Bukkit;
import org.bukkit.Chunk;
import org.bukkit.Location;
import org.bukkit.attribute.Attribute;
import org.bukkit.attribute.AttributeInstance;
import org.bukkit.block.Block;
import org.bukkit.block.data.type.CopperBulb;
import org.bukkit.entity.Player;
import org.bukkit.util.Vector;

import java.util.HashMap;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class SteamPipe {

    public final boolean isDualWay;
    private PipeStatus pipeStatus = PipeStatus.INACTIVE;
    private boolean shuttingDown = false;
    private final List<Location> nodes;
    private final HashMap<Player, Integer> playerEdge = new HashMap<>();

    private int tick = 0;
    private final List<Block> firstBulbs; // dette funker ikke engang pga man må faktisk ta setblockstate for at noesomhelst skal skje!
    private final List<Block> secondBulbs;

    public SteamPipe(boolean isDualWay, List<Location> nodes, Axis entryAxis, Axis exitAxis) {
        this.isDualWay = isDualWay;
        this.nodes = nodes;
        this.firstBulbs = Util.getOrthogonalLocations(nodes.getFirst(), entryAxis).stream()
                .map(Location::getBlock)
                .filter(block -> block.getType().data.equals(CopperBulb.class))
                .collect(Collectors.toList());
        this.secondBulbs = Util.getOrthogonalLocations(nodes.getLast(), exitAxis).stream()
                .map(Location::getBlock)
                .filter(block -> block.getType().data.equals(CopperBulb.class))
                .collect(Collectors.toList());
    }

    // the chunks that has the entry and exit. only check if players are near entry/exit if theyre in these chunks, to save performance
    public Set<Chunk> getTickedChunks() {
        Set<Chunk> chunks = Util.getChunksAround(nodes.getFirst(), 3);
        if (isDualWay) chunks.addAll(Util.getChunksAround(nodes.getLast(), 3));
        return chunks;
    }

    private void setPipeStatus(PipeStatus status) {
        if (this.pipeStatus == status) return;
        switch (status) {
            case ACTIVE -> {
                pipeStatus = status;
                firstBulbs.forEach(bulb -> setLit(bulb,true));
                secondBulbs.forEach(bulb -> setPowered(bulb, true));
            }
            case ACTIVE_REVERSED -> {
                pipeStatus = status;
                firstBulbs.forEach(bulb -> setPowered(bulb, true));
                secondBulbs.forEach(bulb -> setLit(bulb,true));
            }
            case INACTIVE -> {
                shuttingDown = true;

                Bukkit.getScheduler().runTaskLater(Main.getPlugin(), () -> {
                    if (!playerEdge.isEmpty()) return;
                    pipeStatus = status;
                    firstBulbs.forEach(bulb -> setPowered(bulb, false));
                    firstBulbs.forEach(bulb -> setLit(bulb, false));

                    secondBulbs.forEach(bulb -> setPowered(bulb, false));
                    secondBulbs.forEach(bulb -> setLit(bulb, false));
                    shuttingDown = false;
                }, 20L);
            }
        }
    }

    private static void setPowered(Block block, boolean powered) {
        CopperBulb bulb = (CopperBulb) block.getBlockData();
        bulb.setPowered(powered);
        block.setBlockData(bulb);
    }

    private static void setLit(Block block, boolean lit) {
        CopperBulb bulb = (CopperBulb) block.getBlockData();
        bulb.setLit(lit);
        block.setBlockData(bulb);
    }

    public void checkProximity(Player p) {
        if (playerEdge.containsKey(p)) return;
        if (isDualWay) {
            if (pipeStatus != PipeStatus.ACTIVE_REVERSED && checkProximity(true , p)) return;
            if (pipeStatus != PipeStatus.ACTIVE          && checkProximity(false, p)) return;
            return;
        }
        checkProximity(true, p);
    }

    private boolean checkProximity(boolean firstEntry, Player p) {
        Location enterLocation = firstEntry ? nodes.getFirst() : nodes.getLast();
        if (enterLocation.distanceSquared(p.getLocation()) < 9) {
            Vector playerEntryDir = enterLocation.clone().subtract(p.getLocation()).toVector();
            Location nextNode = firstEntry ? nodes.get(1) : nodes.get(nodes.size() - 2);
            Vector pipeEntryDir = nextNode.clone().subtract(enterLocation).toVector();
            if (playerEntryDir.normalize().dot(pipeEntryDir.normalize()) < 0.6) return false; // if youre close to the entry point, but not standing at the front
            // you wont get sucked in. this is to fix getting stuck at the back of the entry
            // 0.6 here means around 55 deg, so if angle between player->entry and entry->nextnode is more than that its not gonna suck the player in

            playerEdge.put(p, firstEntry ? -1 : nodes.size());
            //Bukkit.broadcast(Component.text("Registerd enter"));

            if (firstEntry) setPipeStatus(PipeStatus.ACTIVE);
            else setPipeStatus(PipeStatus.ACTIVE_REVERSED);

            return true;
        }
        return false;
    }

    public void tick() {
        if (pipeStatus == PipeStatus.INACTIVE) return;
        tick++;

        playerEdge.forEach((p, lastNode) -> {
            int nextNode = pipeStatus == PipeStatus.ACTIVE ? lastNode + 1 : lastNode - 1;
            Location currentNodeLoc = (lastNode < 0 || lastNode == nodes.size()) ? p.getLocation() : nodes.get(lastNode);
            Location nextNodeLoc = nodes.get(nextNode);
            Location pLoc = p.getLocation();

            int nodeX = currentNodeLoc.getBlockX();
            int nodeY = currentNodeLoc.getBlockY();
            int nodeZ = currentNodeLoc.getBlockZ();
            if (tick % 5 == 0) { // if the player goes outside of the pipe bounds, they exited
                if ((nodeX == nextNodeLoc.getBlockX() && Math.abs(pLoc.getBlockX() - nodeX) > 1)
                 || (nodeY == nextNodeLoc.getBlockY() && Math.abs(pLoc.getBlockY() - nodeY) > 1)
                 || (nodeZ == nextNodeLoc.getBlockZ() && Math.abs(pLoc.getBlockZ() - nodeZ) > 1)) {
                    exitPlayer(p);
                    return;
                }
            }

            boolean isEntering = pipeStatus == PipeStatus.ACTIVE ? lastNode == -1 : lastNode == nodes.size();
            boolean onFirstNode = pipeStatus == PipeStatus.ACTIVE ? nextNode == 0 : lastNode == nodes.size() - 1;
            boolean onLastNode = pipeStatus == PipeStatus.ACTIVE ? nextNode == nodes.size() - 1 : nextNode == 0;
            if (isEntering) {
                if (nextNodeLoc.clone().distanceSquared(pLoc) > 12) {
                    playerEdge.put(p, -2); // the player exited and will be removed
                }
                AttributeInstance scaleAttribute = p.getAttribute(Attribute.SCALE);
                double scale = scaleAttribute.getBaseValue();
                scale -= 0.05;
                if (scale < 0.4) scale = 0.4;
                scaleAttribute.setBaseValue(scale);
            } else if (onFirstNode) {
                AttributeInstance scaleAttribute = p.getAttribute(Attribute.SCALE);
                scaleAttribute.setBaseValue(0.4);
            }

            Vector a = nextNodeLoc.clone().subtract(currentNodeLoc).toVector().normalize();
            Vector v = p.getVelocity();

            double distanceToNextNode = pLoc.distanceSquared(nextNodeLoc);
            if ((isEntering || v.lengthSquared() > 1)) { // get slowly sucked in, then go fast but dont go too much faster than 1b/t
                a.multiply(new Vector(0.1, 0.3, 0.1));
                if (isEntering) { // the closer you get to the entry, the stronger the pull
                    double multiply = (9 - distanceToNextNode) / 6;
                    a.multiply(new Vector(multiply, 1.5, multiply));
                }
            }
            v.add(a);
            p.setVelocity(v);
            boolean reachedNextNode = isEntering ? distanceToNextNode < 0.5 : distanceToNextNode < 1; // you must be closer to the first node to have reached it
            // this secures that the player is actually inside the pipe
            if (reachedNextNode) {
                if (onLastNode) {
                    exitPlayer(p);
                } else {
                    playerEdge.put(p, nextNode);
                }
            }
        });
        playerEdge.entrySet().removeIf(entry -> entry.getValue() == -2);
        if (playerEdge.isEmpty() && !shuttingDown) setPipeStatus(PipeStatus.INACTIVE);
        if (tick % 4 == 0) {
            updateAnimation();
        }
    }

    void exitPlayer(Player p) {
        playerEdge.put(p, -2); // the player exited and will be removed
        Bukkit.getScheduler().runTaskLater(Main.getPlugin(), () -> p.getAttribute(Attribute.SCALE).setBaseValue(1.0), 2L);
    }

    void updateAnimation() {
        List<Block> entryBulbs = pipeStatus == PipeStatus.ACTIVE ? firstBulbs : secondBulbs;
        List<Block> exitBulbs = pipeStatus == PipeStatus.ACTIVE ? secondBulbs : firstBulbs;

        if (!entryBulbs.isEmpty()) {
            entryBulbs.forEach(bulb -> setPowered(bulb, false));
            int entryBulb = (tick / 4) % entryBulbs.size();
            setPowered(entryBulbs.get(entryBulb), true);
        }
        if (!exitBulbs.isEmpty()) {
            exitBulbs.forEach(bulb -> setLit(bulb,false));
            int exitBulb = (tick / 4) % exitBulbs.size();
            setLit(exitBulbs.get(exitBulb), true);
        }
    }

    private enum PipeStatus {
        INACTIVE,
        ACTIVE,
        ACTIVE_REVERSED
    }
}

File: src/main/java/gruvexp/bbminigames/mechanics/SteamPipe.kt
package gruvexp.bbminigames.mechanics

import gruvexp.bbminigames.Main
import gruvexp.bbminigames.twtClassic.BotBows
import gruvexp.bbminigames.twtClassic.BotBowsPlayer
import gruvexp.bbminigames.twtClassic.effect.PlayerEffectManager
import gruvexp.bbminigames.util.editData
import gruvexp.bbminigames.util.getChunksAround
import gruvexp.bbminigames.util.getOrthogonalLocations
import org.bukkit.Axis
import org.bukkit.Bukkit
import org.bukkit.Chunk
import org.bukkit.Location
import org.bukkit.block.Block
import org.bukkit.block.data.type.CopperBulb
import org.bukkit.util.Vector
import kotlin.math.abs

class SteamPipe(val isDualWay: Boolean, private val nodes: List<Location>, entryAxis: Axis, exitAxis: Axis) {
    private var pipeStatus = PipeStatus.INACTIVE
    private var shuttingDown = false
    private val playerEdge = mutableMapOf<BotBowsPlayer, Int>()

    private var tick = 0
    private val firstBulbs: List<Block> = nodes.first().getOrthogonalLocations(entryAxis)
        .map { it.block }
        .filter { it.type.data == CopperBulb::class.java }
    private val secondBulbs: List<Block> = nodes.last().getOrthogonalLocations(exitAxis)
        .map { it.block }
        .filter { it.type.data == CopperBulb::class.java }

    val tickedChunks: Set<Chunk>
        // the chunks that has the entry and exit. only check if players are near entry/exit if theyre in these chunks, to save performance
        get() {
            val chunks = nodes.first().getChunksAround(3)
            if (isDualWay) chunks.addAll(nodes.last().getChunksAround( 3))
            return chunks
        }

    private fun setPipeStatus(status: PipeStatus) {
        if (pipeStatus == status) return
        when (status) {
            PipeStatus.ACTIVE -> {
                pipeStatus = status
                firstBulbs.forEach { it.editData<CopperBulb> { isLit = true } }
                secondBulbs.forEach { it.editData<CopperBulb> { isPowered = true } }
            }

            PipeStatus.ACTIVE_REVERSED -> {
                pipeStatus = status
                firstBulbs.forEach { it.editData<CopperBulb> { isPowered = true } }
                secondBulbs.forEach { it.editData<CopperBulb> { isLit = true } }
            }

            PipeStatus.INACTIVE -> {
                shuttingDown = true

                Bukkit.getScheduler().runTaskLater(Main.plugin, Runnable {
                    if (playerEdge.isNotEmpty()) return@Runnable
                    pipeStatus = status
                    firstBulbs.forEach { it.editData<CopperBulb> { isPowered = false; isLit = false } }

                    secondBulbs.forEach { it.editData<CopperBulb> { isPowered = false; isLit = false } }
                    shuttingDown = false
                }, 20L)
            }
        }
    }

    fun checkProximity(p: BotBowsPlayer) {
        if (p in playerEdge) return
        if (isDualWay) {
            if (pipeStatus != PipeStatus.ACTIVE_REVERSED && checkProximity(true, p)) return
            if (pipeStatus != PipeStatus.ACTIVE && checkProximity(false, p)) return
            return
        }
        checkProximity(true, p)
    }

    private fun checkProximity(firstEntry: Boolean, bp: BotBowsPlayer): Boolean {
        val enterLocation = if (firstEntry) nodes.first() else nodes.last()
        if (enterLocation.distanceSquared(bp.location) < 9) {
            val playerEntryDir = enterLocation.clone().subtract(bp.location).toVector()
            val nextNode = if (firstEntry) nodes[1] else nodes[nodes.lastIndex - 1]
            val pipeEntryDir = nextNode.clone().subtract(enterLocation).toVector()
            if (playerEntryDir.normalize().dot(pipeEntryDir.normalize()) < 0.6) return false // if youre close to the entry point, but not standing at the front


            // you wont get sucked in. this is to fix getting stuck at the back of the entry
            // 0.6 here means around 55 deg, so if angle between player->entry and entry->nextnode is more than that its not gonna suck the player in
            playerEdge[bp] = if (firstEntry) -1 else nodes.size

            //Bukkit.broadcast(Component.text("Registerd enter"));
            setPipeStatus(if (firstEntry) PipeStatus.ACTIVE else PipeStatus.ACTIVE_REVERSED)

            return true
        }
        return false
    }

    fun tick() {
        if (pipeStatus == PipeStatus.INACTIVE) return
        tick++

        playerEdge.forEach { (bp: BotBowsPlayer, currentNode: Int) ->
            val nextNode = if (pipeStatus == PipeStatus.ACTIVE) currentNode + 1 else currentNode - 1

            val currentNodeLoc = if (currentNode < 0 || currentNode == nodes.size) bp.location else nodes[currentNode]
            val nextNodeLoc = nodes[nextNode]

            val pLoc = bp.location
            val nodeX = currentNodeLoc.blockX
            val nodeY = currentNodeLoc.blockY
            val nodeZ = currentNodeLoc.blockZ
            if (tick % 5 == 0) { // if the player goes outside of the pipe bounds, they exited
                if (nodeX == nextNodeLoc.blockX && abs(pLoc.blockX - nodeX) > 1
                    || nodeY == nextNodeLoc.blockY && abs(pLoc.blockY - nodeY) > 1
                    || nodeZ == nextNodeLoc.blockZ && abs(pLoc.blockZ - nodeZ) > 1
                ) {
                    exitPlayer(bp)
                    return@forEach
                }
            }

            val isEntering = if (pipeStatus == PipeStatus.ACTIVE) currentNode == -1 else currentNode == nodes.size
            val onFirstNode = if (pipeStatus == PipeStatus.ACTIVE) currentNode == 0 else currentNode == nodes.size - 1
            val onLastNode = if (pipeStatus == PipeStatus.ACTIVE) nextNode == nodes.size - 1 else nextNode == 0

            if (isEntering) {
                if (nextNodeLoc.clone().distanceSquared(pLoc) > 12) {
                    playerEdge[bp] = -2 // the player exited and will be removed
                }
                bp.effectManager.applyScale(
                    PlayerEffectManager.ScaleSource.STEAM_PIPE,
                    0.4,
                    PlayerEffectManager.ScalePriority.OVERRIDE,
                    3L
                )
            } else if (onFirstNode) {
                bp.effectManager.applyScale(
                    PlayerEffectManager.ScaleSource.STEAM_PIPE,
                    0.4,
                    PlayerEffectManager.ScalePriority.OVERRIDE,
                    null
                ) // will be small "forever" until exiting
            }

            val a = nextNodeLoc.clone().subtract(currentNodeLoc).toVector().normalize()

            val distanceToNextNode = pLoc.distanceSquared(nextNodeLoc)
            if (isEntering || bp.avatar.entity.velocity.lengthSquared() > 1) { // get slowly sucked in, then go fast but dont go too much faster than 1b/t
                a.multiply(Vector(0.1, 0.3, 0.1))
                if (isEntering) { // the closer you get to the entry, the stronger the pull
                    val multiply = (9 - distanceToNextNode) / 6
                    a.multiply(Vector(multiply, 1.5, multiply))
                }
            }
            bp.avatar.entity.apply { velocity = velocity.add(a) }
            val reachedNextNode = if (isEntering) distanceToNextNode < 0.5 else distanceToNextNode < 1 // you must be closer to the first node to have reached it
            // this secures that the player is actually inside the pipe
            if (reachedNextNode) {
                if (onLastNode) {
                    exitPlayer(bp)
                } else {
                    playerEdge[bp] = nextNode
                }
            }
        }
        playerEdge.entries.removeIf { it.value == -2 }
        if (playerEdge.isEmpty() && !shuttingDown) setPipeStatus(PipeStatus.INACTIVE)
        if (tick % 4 == 0) {
            updateAnimation()
        }
    }

    fun exitPlayer(bp: BotBowsPlayer) {
        playerEdge[bp] = -2 // the player exited and will be removed
        bp.effectManager.applyScale(
            PlayerEffectManager.ScaleSource.STEAM_PIPE,
            0.4,
            PlayerEffectManager.ScalePriority.OVERRIDE,
            3L,
            10
        )
        BotBows.debugMessage("3. will be big in 3 ticks")
    }

    fun updateAnimation() {
        val entryBulbs = if (pipeStatus == PipeStatus.ACTIVE) firstBulbs else secondBulbs
        val exitBulbs = if (pipeStatus == PipeStatus.ACTIVE) secondBulbs else firstBulbs

        if (entryBulbs.isNotEmpty()) {
            entryBulbs.forEach { it.editData<CopperBulb> { isPowered = false } }
            val entryBulb = tick / 4 % entryBulbs.size
            entryBulbs[entryBulb].editData<CopperBulb> { isPowered = true }
        }
        if (exitBulbs.isNotEmpty()) {
            exitBulbs.forEach { it.editData<CopperBulb> { isLit = false } }
            val exitBulb = tick / 4 % exitBulbs.size
            exitBulbs[exitBulb].editData<CopperBulb> { isLit = true }
        }
    }

    private enum class PipeStatus {
        INACTIVE,
        ACTIVE,
        ACTIVE_REVERSED
    }
}
